### PR TITLE
132 make bulk file format configurable

### PR DESCRIPTION
**Pull request by je-ik**

### CODE REVIEW

**Comment by je-ik** [comment]
This closes O2-Czech-Republic#132

**Comment by je-ik** [comment]
This closes O2-Czech-Republic#132

File: direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/FileSystem.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Stream;
+
+/** A proxima's abstraction of bulk FS. */
+@Internal
+public interface FileSystem extends Serializable {
+
+  static FileSystem local(File parent, NamingConvention convention) {
```
Comment (dmvk) [inline direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/FileSystem.java]: Can we use `java.nio` here instead of the old API?

https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html

File: direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/Path.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  static Path local(FileSystem fs, File path) {
+    return new Path() {
+
+      @Override
+      public InputStream reader() throws IOException {
+        return new FileInputStream(path);
+      }
+
+      @Override
+      public OutputStream writer() throws IOException {
```
Comment (dmvk) [inline direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/Path.java]: What if file already exists (we want to make sure files are immutable accross all FS impls) / is directory?

File: direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/Reader.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ * limitations under the License.
+ */
+package cz.o2.proxima.direct.bulk;
+
+import cz.o2.proxima.annotations.Internal;
+import cz.o2.proxima.storage.StreamElement;
+
+/** Proxima's representation of generic reader. */
+@Internal
+public interface Reader extends AutoCloseable, Iterable<StreamElement> {
```
Comment (dmvk) [inline direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/Reader.java]: Is this supposed to be reiterable? If not can we change this to `Iterator` or `Stream`?

File: direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/BinaryBlobFormat.java [hunk 4]
```diff
@@ -116,22 +116,29 @@ public void close() throws IOException {
...
 
-  public static class Reader implements Iterable<StreamElement>, Closeable {
+  public static class BinaryBlobReader implements Reader {
 
+    private final Path path;
     private final Parser<Serialization.Element> parser = Serialization.Element.parser();
     private final EntityDescriptor entity;
     private final Serialization.Header header;
     private final String blobName;
     private DataInputStream blobStream = null;
```
Comment (dmvk) [inline direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/BinaryBlobFormat.java]: final

File: direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/BinaryBlobFormat.java [hunk 5]
```diff
@@ -116,22 +116,29 @@ public void close() throws IOException {
...
     }
+
+    @Override
+    public Path getPath() {
+      return path;
+    }
   }
 
-  public static class Reader implements Iterable<StreamElement>, Closeable {
+  public static class BinaryBlobReader implements Reader {
```
Comment (dmvk) [inline direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/BinaryBlobFormat.java]: nit: I'd prefer `  public static class Reader implements cz.o2.proxima.direct.bulk.Reader {` (same applies to writer)

File: direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/BinaryBlobFormat.java [hunk 6]
```diff
@@ -237,75 +244,62 @@ private StreamElement fromBytes(byte[] data) throws IOException {
     }
 
     @Override
-    public void close() throws IOException {
+    public void close() {
       if (blobStream != null) {
-        blobStream.close();
+        ExceptionUtils.unchecked(blobStream::close);
```
Comment (dmvk) [inline direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/BinaryBlobFormat.java]: nit: Maybe we should have our own version of `Closeables.close(Closeable)` instead. (the guava one rethrows the exception 🤦‍♂ )

File: direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/BinaryBlobFormat.java [hunk 7]
```diff
@@ -237,75 +244,62 @@ private StreamElement fromBytes(byte[] data) throws IOException {
...
-    return new Reader(entity, path.getName(), new FileInputStream(path));
+  @Override
+  public BinaryBlobWriter openWriter(Path path, EntityDescriptor entity) throws IOException {
+    return new BinaryBlobWriter(path, writeGzip, path.writer());
   }
 
-  @Getter private final File path;
+  @Override
+  public String fileSuffix() {
+    return writeGzip ? "blob.gz" : "blob";
```
Comment (dmvk) [inline direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/BinaryBlobFormat.java]: what about other compression formats?

File: direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/DefaultNamingConvention.java [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+        DIR_FORMAT.format(
+            LocalDateTime.ofInstant(
+                Instant.ofEpochMilli(boundary), ZoneId.ofOffset("UTC", ZoneOffset.UTC)));
+    return String.format(
+        "/%s%s-%d_%d_%s.%s",
+        date, prefix, boundary, boundary + rollPeriodMs, uuidGenerator.apply(), suffix);
+  }
+
+  @Override
+  public Set<String> prefixesOf(long startStamp, long endStamp) {
```
Comment (dmvk) [inline direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/DefaultNamingConvention.java]: Would it make sense to change the API to work with Instants instead? (That applies to FileSystem abstraction too). (these are easier to work with and most of the methods I see convert timestamps to Instants anyway)

File: direct/ingest-server/src/main/java/cz/o2/proxima/server/ReplicationController.java [hunk 9]
```diff
@@ -452,6 +452,11 @@ void ingestElement(StreamElement ingest, OnNextContext context) {
                         context::confirm,
                         context::fail));
           }
+
+          @Override
+          public void onIdle(OnIdleContext context) {
+            writer.updateWatermark(context.getWatermark());
```
Comment (dmvk) [inline direct/ingest-server/src/main/java/cz/o2/proxima/server/ReplicationController.java]: how is this change related to this PR?

Comment (dmvk) [reply direct/ingest-server/src/main/java/cz/o2/proxima/server/ReplicationController.java]: oh, I see there is a new `updateWatermark` on the Writer interface 👍

File: direct/core/src/main/java/cz/o2/proxima/direct/core/BulkAttributeWriter.java [hunk 10]
```diff
@@ -52,4 +52,11 @@ default Type getType() {
    * @param statusCallback callback to commit the data
    */
   void write(StreamElement data, long watermark, CommitCallback statusCallback);
+
+  /**
+   * Update watermark when no input data arrives.
```
Comment (dmvk) [inline direct/core/src/main/java/cz/o2/proxima/direct/core/BulkAttributeWriter.java]: I think `when no input data arr...` is just the impl. detail. Should we change this to `Update watermark of this writer.` instead.

File: direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/AbstractBulkFileSystemAttributeWriter.java [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import lombok.Getter;
+import lombok.ToString;
+import lombok.extern.slf4j.Slf4j;
+
+/** An abstract parent class for {@link BulkAttributeWriter}. */
+@Slf4j
+public abstract class AbstractBulkFileSystemAttributeWriter extends AbstractBulkAttributeWriter {
```
Comment (dmvk) [inline direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/AbstractBulkFileSystemAttributeWriter.java]: This is the only implementation of `AbstractBulkAttributeWriter`. Maybe we can merge these?

File: direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/Reader.java (hunk 3 above)
Comment (je-ik) [reply direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/Reader.java]: I think there should be no problem with reiterating this. It is not possible to reiterate it twice at the same time, but there is no reason for that. I'd keep it as it is, because it is @Internal API and it is handy to have it `Iterable`.

File: direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/BinaryBlobFormat.java (hunk 4 above)
Comment (je-ik) [reply direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/BinaryBlobFormat.java]: That's not possible, because `close` method should be idempotent and must therefore set this to `null`.

File: direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/BinaryBlobFormat.java (hunk 5 above)
Comment (je-ik) [reply direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/BinaryBlobFormat.java]: I don't like classes extending other classes with same name, although in different package. I'll keep as is.

File: direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/BinaryBlobFormat.java (hunk 6 above)
Comment (je-ik) [reply direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/BinaryBlobFormat.java]: I think this is readable enough. The quava version might be `Closeables.closeQuietly(Closeable)`, but that is nearly as verebose as this.

File: direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/BinaryBlobFormat.java (hunk 7 above)
Comment (je-ik) [reply direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/BinaryBlobFormat.java]: Nothing for now, adding more compression format is out of scope of this PR. It would mean adding `CompressionCodec` abstraction. We might do it in the future.

File: direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/DefaultNamingConvention.java (hunk 8 above)
Comment (je-ik) [reply direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/DefaultNamingConvention.java]: This is actually a big topic. :-)
I think `Instant` and `Duration` should be treated in similar way to `Optional` - i.e. use it in public APIs, convert to raw types in internal APIs. This is internal API and I would keep it as is for now. We can reiterate on this later.

File: direct/core/src/main/java/cz/o2/proxima/direct/core/BulkAttributeWriter.java (hunk 10 above)
Comment (je-ik) [reply direct/core/src/main/java/cz/o2/proxima/direct/core/BulkAttributeWriter.java]: I don't think this is implementation detail, as the watermark is passed to `write` method. This is called *solely* when there is no input data currently.

File: direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/AbstractBulkFileSystemAttributeWriter.java (hunk 11 above)
Comment (je-ik) [reply direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/AbstractBulkFileSystemAttributeWriter.java]: Hm, I think I can move the one from `direct-core` here and merge it. :+1:

File: direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/FileSystem.java (hunk 1 above)
Comment (je-ik) [reply direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/FileSystem.java]: I guess we can. The "problem" would be, that JUnit 4 and its TemporaryFolder returns the old API. I would leave this refactor as part of moving to JUnit 5, which can create temporary files using the new API. Otherwise we would just convert APIs back and forth, without any significant benefit.

File: direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/Path.java (hunk 2 above)
Comment (je-ik) [reply direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/Path.java]: Then this should throw exception, do you see issues with that?

File: direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/AbstractBulkFileSystemAttributeWriter.java (hunk 11 above)
Comment (je-ik) [reply direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/AbstractBulkFileSystemAttributeWriter.java]: Hm, this is not the only implementation. There is `InMemStorage.Writer` implementation. Therefore, I'd keep it separated, because I don't want to pull this to core and core cannot depend on this (cyclic dependency).

File: direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/Utils.java [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+import java.net.InetAddress;
+import java.nio.charset.Charset;
+import java.security.MessageDigest;
+import java.time.Duration;
+import java.util.Base64;
+import java.util.Map;
+import java.util.Optional;
+
+/** Utility class to be shared between various bulk storages. */
+public class Utils {
```
Comment (dmvk) [inline direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/Utils.java]: We should split this class into `NamingConvetions` and `FileFormats`

File: direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/Utils.java [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+          .map(cls -> Classpath.newInstance(cls, NamingConvention.class))
+          .orElse(
+              NamingConvention.defaultConvention(
+                  Duration.ofMillis(rollPeriodMs), prefix, format.fileSuffix()));
+    } catch (Exception ex) {
+      throw new RuntimeException(ex);
+    }
+  }
+
+  public static FileFormat getFileFormat(String cfgPrefix, Map<String, Object> cfg) {
```
Comment (dmvk) [inline direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/Utils.java]: javadocs?
